[FIX] website: consider other media types for Social Media snippet icons

sobo-odoo opened the pull request:
[FIX] website: consider other media types for Social Media snippet icons

When using the "Social Media" snippet, there are two situations leading to a traceback. Here are the steps:

1) First issue
- In edit mode, drop the "Social Media" snippet.
- Add a new social network.
- Replace the icon by a real image (so not by an other icon).
- Set the URL of the new social network to a "relevant" one (e.g. google, facebook).
=> traceback

2) Second issue
- Drop "Social Media" and change the list in order to have the first social network with a real image as icon: either replace the first icon with an image, or add a custom one with an image an move it to the top.
- Add a new social network. 
=> traceback

Both issues happen because only icons set as `<i>` elements are taken into account when modifying a social network (`querySelector('i')`) and this element may not exist if the icon had another tag (like an image or a span or even a video (as the media dialog allows to upload them)). As there is no check ensuring that the element exists before changing the different social media classes, there is a traceback when the `<i>` is not found.

For the first issue, it happens because when changing the URL, we first check if it matches with relevant ones or if an icon matching it exists. If it is the case, we then update the social media classes to set the matching icon => it can only work if the icon is a `<i>`.

The second issue happens because when adding a new social network, it actually copies the style of the first one (unless there are none, a new one is created in that case). If the first one is an image, it fails when updating the classes because it is not a `<i>`.

This commit fixes these issues by simply adding checks ensuring the icon element exists before trying to modify it. It also adds tests, ensuring these use cases work as expected.

opw-3538230

---
[FIX] website: consider `<span>` icons in Social Media snippet

In the Social Media snippet options, only icons with a `<i>` tag are
considered when updating the social media classes (when adding a new
social network by copying the first one and when replacing the icon by
another one matching the URL). However, there are cases where the icon
is a `<span>` element (with a `fa` class) and not a `<i>`.

For example:
- Replace a social network icon by a real image.
- Re-replace this image by an icon.
=> the icon element tag is `<span>` and not `<i>`.

In this case, after setting its style correctly (e.g. adding the round
shape) to make it look like the other icons, we notice that it does not
behave like the other ones:
- It is not aligned with the other icons, because the CSS rule aligning
them only targets `<i>`.
- Its icon is not replaced by a matching one when changing the URL with
a "relevant" one (e.g. google, facebook).

This commit considers `<span>` icons in the Social Media options, in
addition to `<i>` icons, in order for them to all behave the same way.

opw-3538230

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/142280).

sobo-odoo commented:
@qsm-odoo @rdeodoo ready for review :slightly_smiling_face: 

FYI the second commit is not related to the ticket => I found it when testing for a bit and thought it could be nice to fix it too :slightly_smiling_face:  I am not sure if it is really ok for stable because of the CCS rule change so tell me and I will drop it if it's not ok

Also, when discussing with RDE, we think that maybe the image upload with the media dialog when we are in the Social Media snippet should be improved (in master and in an other task):
- replace an icon by an image => the image takes the whole snippet width. We can set its size to 25% to make it smaller but it is sometimes not small enough and the `<a>` element still takes the whole space so we can't put the icon next to the others (they go one above another). Of course, maybe the user should upload a better image that would have an appropriate size but it still does not align properly :grimacing:  It could be nice to add some code to make the image look like a real icon, so resized the same way as the icons: for example, the Social Media "Size" option does not apply to images so it could be cool to improve that
- to avoid the `<span>` issue my second commit fixes, maybe the added icon should always be a `<i>`

A more drastic way that would fix all that could be to only allow icons in the media dialog (currently, we can upload anything, even videos). But after thinking about it, the images must stay allowed, or else we could not upload "customized" logo (that would go better with the website style for example).

What do you think about that ? If you agree, I will create a task :slightly_smiling_face:

sobo-odoo commented:
@qsm-odoo @rdeodoo re-pinging you in case you didn't see (since it is a bugfix) :slightly_smiling_face:

qsm-odoo commented on addons/website/static/src/snippets/s_social_media/options.js:
```diff
@@ -146,13 +149,16 @@ options.registry.SocialMedia = options.Class.extend({
-                            anchorEl.classList.add(`s_social_media_${socialMedia}`);
-                            iEl.classList.add(`fa-${socialMedia}`);
+                            if (iEl) {
+                                anchorEl.classList.add(`s_social_media_${socialMedia}`);
```
I am curious:
- Why only adding this class if there is an icon?
- Already like that before butu: why calling `_removeSocialMediaClasses` only if we found a social media?

sobo-odoo replied:
> Why only adding this class if there is an icon

From what I saw, the `s_social_media_...` class only purpose is to set the icon color to the one matching the social media used (like the facebook blue, or google red) so it will not break anything on elements other than icons but it will be useless (so better not use it) 
![image](https://github.com/odoo/odoo/assets/102796037/45dc0070-26c0-4287-a943-c36ab9d85711)

> why calling _removeSocialMediaClasses

This function is used to clean the previous social media classes before setting the new ones to avoid "clashes" between them when setting the new matching classes (so remove anything "facebook-related" before adding the "google-specific") given that the `fa` classes manage the icon used in the "icon" (and `s_social_media_...` the color as said previously)

Note that this function is also called when adding a new icon (above, before the first diff), to clean it since it is a copy of the first one  
But also note that replacing the icon (so with the media dialog) will not remove the `s_social_media_` class from the anchor element, so I think we remove the classes only if really necessary :slightly_smiling_face:

qsm-odoo replied:
> so it will not break anything on elements other than icons but it will be useless (so better not use it)

I have the opposite opinion: the fact it is useless in some configuration is not important. What is better IMO is that if the option detects it is a facebook URL, the related class is added, always, whatever its effect is. A custo could take profit of that (and actually, as we are in stable, best change the behavior as less as possible) :+1: 

> to clean the previous social media classes before setting the new ones

Yes, but since you now were seting the new one only `if (iEl)`, it seems weird: either cleaning should be in that `if` too or outside of the main one I think :wink: But as I said above, I think we should rather add the  `s_social_media_XXX` class as soon as we find a social media :+1:

qsm-odoo commented on addons/website/static/src/snippets/s_social_media/options.js:
```diff
@@ -122,7 +122,7 @@ options.registry.SocialMedia = options.Class.extend({
                         this._removeSocialMediaClasses(anchorEl);
                     }
-                    const iEl = anchorEl.querySelector("i");
+                    const iEl = anchorEl.querySelector("i, span.fa");
```
In 16.0 forward-port, probably adapt to use this: https://github.com/odoo/odoo/commit/4e33b5e4956da53558b0a36ae2e6da13220290ab#diff-d9bb92e950cfb526294e3c97c6c21879e5b2def8b7b70ddeb23960ef254d396bR1364

qsm-odoo reviewed (changes requested):
Thanks for the commits, good explanations :+1: A small question in the review + a small request here: can you adapt the `snippet_social_media` test with a few more steps to test those fixes? :pray:

sobo-odoo commented:
@qsm-odoo I made the changes :slightly_smiling_face: I added additional steps to the test and I moved the "add social media class" part out of the condition checking if it is an icon ([see diff ](https://github.com/odoo/odoo/compare/7defc54b9e12b8ea9dc17165d6fe4b8df30f523a..a9f48f742477fdf49486f38265ebb40502f59bb1) )

Also, just in case you didn't see, you didn't answer to my comment [here](https://github.com/odoo/odoo/pull/142280#issuecomment-1814362105)

qsm-odoo reviewed (approved):
> I added additional steps to the test and I moved the "add social media class" part out of the condition checking if it is an icon

Nice :+1: 

> Also, just in case you didn't see, you didn't answer to my comment

I indeed did not see it / forgot about it :wink: 

> the second commit is not related to the ticket => I am not sure if it is really ok for stable because of the CSS rule change so tell me

It's of course ok: if you are only *adding* CSS behaviors, it's often ok. Here: value > risk.

> we think that maybe the image upload with the media dialog when we are in the Social Media snippet should be improved

You can of course create a task for master :+1: Put it in backlog - green, it's not a high priority.

> to avoid the `<span>` issue my second commit fixes, maybe the added icon should always be a `<i>`

Maybe, but we can forget about this for now.


Thanks for the fixes :+1: 

@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@sobo-odoo @qsm-odoo staging failed: ci/runbot on 70617de6b185b136d8ad6c65653b55277702d05e (view more at https://runbot.odoo.com/runbot/build/54103784)

rdeodoo commented:
@robodoo retry

robodoo commented:
@sobo-odoo @qsm-odoo staging failed: ci/runbot (view more at https://runbot.odoo.com/runbot/build/54130761)

sobo-odoo commented:
@qsm-odoo or @rdeodoo I added a step in the tour, to save the editor at the end, could you re-r+ ? :slightly_smiling_face:  I didn't add it before because there wasn't one to begin with, but the tests don't seem to fail anymore with it :shrug:

qsm-odoo reviewed (approved):
@robodoo rebase-ff r+

Saving the editor at the end of the tour allows to prevent some (newly added) images in the tour to still be loading when the tour ends. It might be the cause of the above fails indeed, thanks @rdeodoo for the hint :+1:

robodoo commented:
Merge method set to rebase and fast-forward.
